src/debug.rs

```diff
@@ -0,0 +1,89 @@
+//! Interacting with debugging agent
+//!
+//! # Example
+//!
+//! This example will show how to terminate the QEMU session. The program
+//! should be running under QEMU with semihosting enabled
+//! (use `-semihosting` flag).
+//!
+//! Target program:
+//!
+//! ```
+//! #[macro_use]
+//! extern crate cortex_m_semihosting;
+//! use cortex_m_semihosting::debug;
+//!
+//! fn main() {
+//!     if 2 == 2 {
+//!         // report success
+//!         debug::exit(0);
+//!     } else {
+//!         // report failure
+//!         debug::exit(1);
+//!     }
+//! }
+//!
+
+/// This values are taken from section 5.5.2 of
+/// "ADS Debug Target Guide" (DUI0058)
+pub enum Exception {
+    // Hardware reason codes
+    BranchThroughZero = 0x20000,
+    UndefinedInstr = 0x20001,
+    SoftwareInterrupt = 0x20002,
+    PrefetchAbort = 0x20003,
+    DataAbort = 0x20004,
+    AddressException = 0x20005,
+    IRQ = 0x20006,
+    FIQ = 0x20007,
+    // Software reason codes
+    BreakPoint = 0x20020,
+    WatchPoint = 0x20021,
+    StepComplete = 0x20022,
+    RunTimeErrorUnknown = 0x20023,
+    InternalError = 0x20024,
+    UserInterruption = 0x20025,
+    ApplicationExit = 0x20026,
+    StackOverflow = 0x20027,
+    DivisionByZero = 0x20028,
+    OSSpecific = 0x20029,
+}
+
+/// Reports to the debugger that the execution has completed.
+///
+/// If `status` is not 0 then an error is reported.
+///
+/// This call can be used to terminate QEMU session, and report back success
+/// or failure.
+///
+/// This call should not return. However, it is possible for the debugger
+/// to request that the application continue. In that case this call
+/// returns normally.
+///
+pub fn exit(status: i8) {
```

Comment:
Sorry for not addressing this earlier but I think the argument here is misleading because, AFAICT, there's no way to actually pass the exit code to the debugger. I'm wondering if we should just take a boolean as the argument or maybe `Result<(), ()>` (`exit(Ok(()))` or `exit(Err(()))`)?